[ao] Fixing obs insertion through dtype propagation

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73274

Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

HDCharles committed:
[ao] Fixing obs insertion through dtype propagation

Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -125,7 +128,7 @@ def node_arg_is_bias(node: Node, arg: Any) -> bool:
     arg: Argument,
     node: Node,
-    node_name_to_target_dtype: Dict[str, Dict[str, Optional[torch.dtype]]],
+    node_name_to_target_dtype: Dict[str, Dict[str, Optional[Union[torch.dtype, type]]]],
```
I feel we probably want to define a type that can represent both Tensor (with dtypes) and primitive types so that they can be easily identified (maybe later in a separate PR)

HDCharles replied:
Yeah getting the type hints to work for this was a bit of a pain, that's a good idea

jerryzh168 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -455,6 +456,78 @@ def all_node_args_have_no_tensors(node: Node, modules: Dict[str, torch.nn.Module
+    """
+    Returns all node arg indices after first
+    """
+    return [i for i, arg in enumerate(node.args)][1:]
```
nit: list(range(1, len(node.args))?

jerryzh168 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -455,6 +456,78 @@ def all_node_args_have_no_tensors(node: Node, modules: Dict[str, torch.nn.Module
+    that are valid for node.args
+    """
+    def arg_indices_func(node: Node) -> List[int]:
+        result = []
+        for i in arg_indices:
+            if i < len(node.args):
+                result.append(i)
+        return result
```
nit: this can be simplified a bit as well I think, like:
```
return [i for i in arg_indices if i < len(node.args)]
```

jerryzh168 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -455,6 +456,78 @@ def all_node_args_have_no_tensors(node: Node, modules: Dict[str, torch.nn.Module
+# so that they can be propagated correctly since inserting observers
+# for them would cause errors
+
+NON_OBSERVABLE_ARG_DICT: Dict[NodeInfo, Dict[Union[type, torch.dtype], Callable[[Node], List[int]]]] = {
```
wondering why don't we have a map from index to arg type? I feel that is probably a bit more intuitive?

HDCharles replied:
I considered something like that but I wasn't sure whether there was a more efficient way to do that than inserting each index to a dict one at a time. This way I create a dict, the key is the dtype and the I insert a single list which seemed more efficient for both creation and searching.

jerryzh168 replied:
I see, if we want to support all the ops we may need to get this information in somewhere instead of handwritten them I think, I'll ask around later about this info to support this generally.

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -821,11 +824,26 @@ def propagate_dtypes_for_known_nodes(
+
+                # when an argument is a tuple, it does not show up as another node so we need to go through
+                # all elements of the tuple manually
+                if isinstance(arg, tuple):
+                    arg_list = list(arg)
+                else:
+                    arg_list = [arg]
```
`arg` could be a nested tuple/list/dict right?

HDCharles replied:
I didn't realize it could be a dict, I'll add a test case for this and make sure to handle it

HDCharles replied:
as far as I can tell it looks like it turns every key/value pair in a dict into its own node. I added a test case where I made a module that accepts a dict as an argument to test this though

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2423,6 +2423,120 @@ def forward(self, x):
+    def _check_not_observed(self, model: nn.Module, node_name_list: List[str]):
+        # check that no obsersvers have the named nodes as an input
+        for node in model.graph.nodes:
+            if hasattr(model, node.name) and isinstance(
```
can we just check the number of observers instead? and also we can make sure the model runs

HDCharles replied:
Can you clarify these comments? The important part is whether any of the named nodes got observed, checking the number of observers doesn't really check the same thing. Its definitely a good idea to see if the model runs though, I'll add that.

jerryzh168 replied:
`node.name` is not a reliable thing, fx is free to change it, so the test may fail when fx tries to change it, since we are hardcoding the node name here

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2423,6 +2423,120 @@ def forward(self, x):
+                for arg in node.args:
+                    self.assertTrue(arg.name not in node_name_list)
+
+    class _non_referece_test_model(nn.Module):
```
what does `non_reference` mean here?

HDCharles replied:
not sure, I'll change it

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2423,6 +2423,120 @@ def forward(self, x):
+            x = self.lin(x)
+            return x
+
+    def _test_dtype_propagation(self, model, *args):
```
I guess it is trying to verify that `y` and `z` are not observed, since we know the these arguments are not Tensors based on non_observable_arg_list, but would be better if we can directly document these things

HDCharles replied:
Ya i'll add a comment

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2542,6 +2542,156 @@ def forward(self, x):
+                for arg in node.args:
+                    self.assertTrue(arg.name not in node_name_list)
+
+    class _non_referece_test_model(nn.Module):
```
nit: upper case for class name

HDCharles replied:
will do

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2542,6 +2542,156 @@ def forward(self, x):
+    def _check_not_observed(self, model: nn.Module, node_name_list: List[str]):
+        # check that no obsersvers have the named nodes as an input
+        for node in model.graph.nodes:
+            if hasattr(model, node.name) and isinstance(
```
looks like we are looking for observers here, I think a better way to do it would be to check:

```
named_modules = model.named_modules()
if node.op == "call_module" and is_activation_post_process(named_modules[node.target]):
    ...
```
also in the checking we can rely on the structure of the graph, instead of node name, e.g. we can check if the value that is observed is the 3rd argument of some op etc.

HDCharles replied:
can you clarify what you mean by rely on structure of the graph? Are you saying I should search for the op in question (i.e. view) and then go through the non_tensor args and see if the source of any of those are an activation_post_process?

jerryzh168 reviewed (changes requested):
looks good, please make the test code more robust, had a few comments inline

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2543,6 +2547,232 @@ def forward(self, x):
+        # indices of nodes and checks whether they are observed
+        def _check_node_indices_not_observed(model, node, indices):
+            for index in indices:
+                if hasattr(node, "args"):
```
this looks a bit weird, is it possible to remove this check?

HDCharles replied:
no, otherwise it fails with tuple arg types

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2543,6 +2547,232 @@ def forward(self, x):
+
+        # this is a helper function (for easier recursion) that looks at specific
+        # indices of nodes and checks whether they are observed
+        def _check_node_indices_not_observed(model, node, indices):
```
does indicies mean different things depending on the type of `node`? shouldn't it always mean node.args[index]?

HDCharles replied:
if an arg is something like x.reshape((-1,x)) the arg shows up as a tuple rather than a node, so you have to index by node[index] rather than node.args[index]

jerryzh168 replied:
you mean https://pytorch.org/docs/stable/generated/torch.Tensor.reshape.html#torch.Tensor.reshape, in this case the index for reshape should just be 0 I think? and we can recurse into the tuple to check individual shape dimensions?

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2587,6 +2591,234 @@ def forward(self, x):
+
+        # this is a helper function (for easier recursion) that checks whether
+        # arg_node is observed
+        def _check_node_indices_not_observed(model, arg_node, node):
```
great! we can remove `_indicies` from the name now I think

jerryzh168 reviewed (approved):
looks good, thanks for addressing the comments!

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
and related to https://github.com/pytorch/pytorch/issues/72698 when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_masked_fill_nontensor_args_not_observed
...
python test/test_quantization.py test_<op>_nontensor_args_not_observed

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34410122](https://our.internmc.facebook.com/intern/diff/D34410122)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34410122).
